File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root) {
        let cnt = 0;
        let max = root.val;
        function dfs(root, maxSoFar){
            if(!root) return;
            if(root.val >= maxSoFar) cnt++; 
            maxSoFar = Math.max(root.val, maxSoFar);
            dfs(root.left, maxSoFar);
            dfs(root.right, maxSoFar);

        }
        dfs(root, max);
        return cnt;      
    }
}

File: Data Structures & Algorithms/binary-tree-longest-consecutive-sequence/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int dfs(TreeNode* root, int curVal, int seqLen){
        if (!root) return seqLen;
        if (root->val == curVal+1) {
            return max(dfs(root->left, root->val, seqLen+1), dfs(root->right, root->val, seqLen+1));
        }
        return max(dfs(root->left, root->val, seqLen), dfs(root->right, root->val, seqLen));
    }

    int longestConsecutive(TreeNode* root) {
        return dfs(root, root->val, 1);
    }
};
